### PR TITLE
Realtime user activity msg subscribing

### PR DESCRIPTION
**Pull request by D3ryk**
* WebSocketing between online users clients
* Sending of messages on three events:

Login ---> triggered after login into application or page has been successfully loaded
Logout ---> triggered after logout from application or page has been successfully unloaded (clossing browser)
Participants ---> once page is loaded user is asking for list of all online users

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -80,6 +80,9 @@ dependencies {
     compile('org.springframework.boot:spring-boot-starter-social-twitter')
     compile('org.springframework.boot:spring-boot-starter-thymeleaf')
     compile('org.springframework.boot:spring-boot-starter-websocket')
+    compile("org.webjars:webjars-locator")
```
Comment (sparkoo) [inline build.gradle]: frontend zavistlosti jdou pres bower (https://github.com/sparkoo/speakup/blob/devel/bower.json). ale to nech, to pak upravim ja, jen pro info

File: src/main/java/cz/sparko/speakup/conf/WebSocketConfig.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    public PresenceEventListener presenceEventListener(SimpMessagingTemplate messagingTemplate) {
+        PresenceEventListener presence = new PresenceEventListener(messagingTemplate, webSocketUserSessionRepository());
+        presence.setLoginEndpoint("/messaging/login");
+        presence.setLogoutEndpoint("/messaging/logout");
+        return presence;
+    }
+
+    @Bean
+    public WebSocketUserSessionRepository webSocketUserSessionRepository() {
+        return new WebSocketUserSessionRepository(new ArrayList<User>());
```
Comment (sparkoo) [inline src/main/java/cz/sparko/speakup/conf/WebSocketConfig.java]: <User> neni potreba, jde pouzit jen diamond <>

File: src/main/java/cz/sparko/speakup/conf/WebSocketConfig.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
+package cz.sparko.speakup.conf;
+
+import cz.sparko.speakup.entities.User;
```
Comment (sparkoo) [inline src/main/java/cz/sparko/speakup/conf/WebSocketConfig.java]: idea mi hlasi nekolik nevyuzitych importu

File: src/main/java/cz/sparko/speakup/events/LoginEvent.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package cz.sparko.speakup.events;
+
+import org.springframework.web.socket.messaging.SessionConnectEvent;
+
+public class LoginEvent {
+    private SessionConnectEvent sessionConnectEvent;
```
Comment (sparkoo) [inline src/main/java/cz/sparko/speakup/events/LoginEvent.java]: final

File: src/main/java/cz/sparko/speakup/controllers/ChatController.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+import org.springframework.messaging.simp.annotation.SubscribeMapping;
+import org.springframework.stereotype.Controller;
+
+import java.util.Collection;
+
+@Controller
+public class ChatController {
+
+    @Autowired
+    private UserSessionRepository userSessionRepository;
```
Comment (sparkoo) [inline src/main/java/cz/sparko/speakup/controllers/ChatController.java]: pouzival bych radsi vsude DI pres konstruktor, tj tady oddelat Autowired a udelat konstruktor
`@Autowired public ChatController(...){this.use....=...}`

bude se to pak jednodusseji mockovat v testech -> nebudes zavislej na spring anotacich a muzes si tam hodit co chces timpadem pro jednoduchej unit test nemusis startovat celej spring context.

File: src/main/java/cz/sparko/speakup/events/LogoutEvent.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package cz.sparko.speakup.events;
+
+import org.springframework.web.socket.messaging.SessionDisconnectEvent;
+
+public class LogoutEvent {
+    private SessionDisconnectEvent sessionDisconnectEvent;
```
Comment (sparkoo) [inline src/main/java/cz/sparko/speakup/events/LogoutEvent.java]: final

File: src/main/java/cz/sparko/speakup/events/listeners/PresenceEventListener.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+import cz.sparko.speakup.events.LogoutEvent;
+import cz.sparko.speakup.repositories.impl.WebSocketUserSessionRepository;
+import org.springframework.context.event.EventListener;
+import org.springframework.messaging.simp.SimpMessagingTemplate;
+import org.springframework.web.socket.messaging.SessionConnectEvent;
+import org.springframework.web.socket.messaging.SessionDisconnectEvent;
+
+public class PresenceEventListener {
+
+    private WebSocketUserSessionRepository webSocketUserSessionRepository;
```
Comment (sparkoo) [inline src/main/java/cz/sparko/speakup/events/listeners/PresenceEventListener.java]: co jde tak hod pls final

File: src/main/java/cz/sparko/speakup/repositories/UserSessionRepository.java [hunk 8]
```diff
@@ -0,0 +1,9 @@
+package cz.sparko.speakup.repositories;
+
+import cz.sparko.speakup.entities.User;
+
+import java.util.Collection;
+
+public interface UserSessionRepository {
```
Comment (sparkoo) [inline src/main/java/cz/sparko/speakup/repositories/UserSessionRepository.java]: public by nemelo byt treba

File: src/main/java/cz/sparko/speakup/repositories/impl/WebSocketUserSessionRepository.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+import cz.sparko.speakup.entities.User;
+import cz.sparko.speakup.events.LoginEvent;
+import cz.sparko.speakup.events.LogoutEvent;
+import cz.sparko.speakup.repositories.UserSessionRepository;
+
+import java.util.Collection;
+
+public class WebSocketUserSessionRepository implements UserSessionRepository {
+
+    private Collection<User> usersSessions;
```
Comment (sparkoo) [inline src/main/java/cz/sparko/speakup/repositories/impl/WebSocketUserSessionRepository.java]: muze byt final

**Comment by D3ryk** [comment]
Ok, dneska po práci na to mrknu a pofixuju co půjde. Potom to mergnu.

File: src/main/java/cz/sparko/speakup/repositories/UserSessionRepository.java (hunk 8 above)
Comment (D3ryk) [inline src/main/java/cz/sparko/speakup/repositories/UserSessionRepository.java]: Tady to nepůjde, protože implementace je v jiným package a potom na sebe nevidí.

File: src/main/java/cz/sparko/speakup/repositories/impl/WebSocketUserSessionRepository.java (hunk 9 above)
Comment (D3ryk) [inline src/main/java/cz/sparko/speakup/repositories/impl/WebSocketUserSessionRepository.java]: Done

File: src/main/java/cz/sparko/speakup/events/listeners/PresenceEventListener.java (hunk 7 above)
Comment (D3ryk) [inline src/main/java/cz/sparko/speakup/events/listeners/PresenceEventListener.java]: Done

File: src/main/java/cz/sparko/speakup/events/LogoutEvent.java (hunk 6 above)
Comment (D3ryk) [inline src/main/java/cz/sparko/speakup/events/LogoutEvent.java]: Done

File: src/main/java/cz/sparko/speakup/events/LoginEvent.java (hunk 4 above)
Comment (D3ryk) [inline src/main/java/cz/sparko/speakup/events/LoginEvent.java]: Done

File: src/main/java/cz/sparko/speakup/conf/WebSocketConfig.java (hunk 2 above)
Comment (D3ryk) [inline src/main/java/cz/sparko/speakup/conf/WebSocketConfig.java]: Done

File: src/main/java/cz/sparko/speakup/conf/WebSocketConfig.java (hunk 3 above)
Comment (D3ryk) [inline src/main/java/cz/sparko/speakup/conf/WebSocketConfig.java]: Done

File: src/main/java/cz/sparko/speakup/controllers/ChatController.java (hunk 5 above)
Comment (D3ryk) [inline src/main/java/cz/sparko/speakup/controllers/ChatController.java]: Done

**Pull request merged**